[RF-DOCS] Update Threading and Code Execution Guide [ci skip] (#3)
OughtPuts/rails: guides/source/threading_and_code_execution.md

After reading this guide, you will know:

> * Where to find automatically concurrent code execution in Rails

Reviewer: "automatically concurrent code" reads odd, maybe just:

```suggestion
* Where to find concurrent code execution in Rails
```

* How to integrate manual concurrency within Rails
* How to wrap all application code using the Rails Executor
* How to affect application reloading

--------------------------------------------------------------------------------

> In-Built Concurrency in Rails

Reviewer 2: The other guides seem to prefer "Built-in" over "In-Built" (there are some other occurences in this guide as well.
```suggestion
Built-in Concurrency in Rails
```

-----------------------------

Rails automatically allows various operations to be performed at the same time
(concurrently) in order for an application to run more efficiently. In this
section, we will explore some of the ways this happens behind the scenes.

When using a threaded web server (such as Rails' default server, Puma) multiple
HTTP requests will be served simultaneously as each request is given its own
controller instance.

Threaded Active Job adapters, including the built-in Async adapter, will
likewise execute several jobs at the same time. Action Cable channels are
managed this way too.

Asynchronous Active Record queries are also performed in the background,
allowing other processes to run on the main thread.

The above mechanisms all involve multiple threads, often managing work for a
unique instance of some object (controller, job, channel), while sharing the
global process space (such as classes and their configurations, and global
variables). As long as the code on each thread doesn't modify any of those

> shared things, the other threads are mostly irrelevant to it.

Reviewer: would this a better ending to that sentence?

```suggestion
variables). As long as the code on each thread doesn't modify any of those
shared things, multiple threads can safely run concurrenlty.
```

Rails' in-built concurrency will cover the day-to-day needs of many application
developers, and ensure applications remain generally performant.

NOTE: You can read more about how to configure Rails' in-built concurrency in

> the [Framework Behavior](#framework-behavior) section.

Reviewer: cleaner without repeating "in-built"

```suggestion
NOTE: You can read more about how to configure Rails concurrency in
the [Framework Behavior](#framework-behavior) section.
```

### `CurrentAttributes` and Threading

The
[`ActiveSupport::CurrentAttributes`](https://edgeapi.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html)
class is a special class in Rails that helps you manage temporary data for each
request in your app, and helps make sure this data is available to the whole
system. It keeps this data separate for every request (even if there are
multiple threads running) and makes sure the data is cleaned up automatically
when the request is done.

You can think of this class as a place to store data that you need to access
anywhere in your app without having to pass it around in your code.

To use the `Current` class to store data, first you need to create a file as

> below, with `attribute` values for the attributes and models whose values you

Reviewer: ```suggestion
To use the `Current` class to store data, first you need to create a file as shown in the example
below, with `attribute` values for the attributes and models whose values you
```

would like to access throughout your application. You can also define a method
(e.g the `user` method below) which, when called, will contain set values:

```ruby
# app/models/current.rb
class Current < ActiveSupport::CurrentAttributes
  attribute :account, :user

  resets { Time.zone = nil }

  def user=(user)
    super
    self.account = user.account
    Time.zone    = user.time_zone
  end
end
```

You will now have access to `Current.user` elsewhere in your application. For
example, when authenticating a user:

```ruby
# app/controllers/concerns/authentication.rb
module Authentication
  extend ActiveSupport::Concern

  included do
    before_action :authenticate
  end

  private
    def authenticate
      if authenticated_user = User.find_by(id: cookies.encrypted[:user_id])

>         Current.user = authenticated_user

Reviewer: NIce. This example ties in nicely with the Authentication documentation.

      else
        redirect_to new_session_url
      end
    end
end
```

WARNING: It’s easy to put too many attributes in the `Current` class and tangle

> your model as a result. Current should only be used for a few, top-level

Reviewer: ```suggestion
your model as a result. `Current` should only be used for a few, top-level
```

globals, like account, user, and request details.

### Isolated Execution State

The `active_support.isolation_level` value in your `configuration.rb` file
provides you the option to define where Rails internal state should be stored

> while tasks are run. If you use a fiber-based server or job processor (e.g.

Reviewer: Would it be in the scope of this guide to explain what a "fiber-based server" is? And what threads are for that matter? If so, something along these line can be added, somewhere in this guide:

_On a single computer, multiple processes can run. Within a single process, multiple threads can run. Within a single thread, multiple fibers can run._

Author: Hm yeah good thought. I wonder if there are other 'going further' type guides for which we need to address this question too. How much are these designed to be 'top level extension' guides, and how much should they explain the concepts. I'll address the rest of the changes first (because we're on a deadline) and put it to Amanda about the explanations - I think we could probably put it on the pile as a to-do to add some beginner level explanation in the first section?

[`falcon`](https://github.com/socketry/falcon)), you should set this value to
`:fiber`, otherwise it is best to set it to `:thread`.

> ### Going Futher

Reviewer: Perhaps that last lead-in sentence doesn't need a subheading?

```suggestion
```

The next section of this guide details advanced ways of wrapping code to ensure
thread safety, and how extensions and applications with particular concurrency
requirements, such as library maintainers, should do this.

Wrapping Application Code
-------------------------

### The Rails Executor

The Rails Executor inherits from the
[`ActiveSupport::ExecutionWrapper`](https://api.rubyonrails.org/classes/ActiveSupport/ExecutionWrapper.html).
The Executor separates application code from framework code by wrapping code
that you've written and is necessary when threads are being used.

The Executor consists of two callbacks: `to_run` and `to_complete`. The `to_run`
callback is called before the application code, and the `to_complete` callback
is called after.

> #### Callbacks

Reviewer: Maybe move this subheading before line 129? as that's talking about callbacks as well.

In a default Rails application, the Rails Executor callbacks are used to:

* track which threads are in safe positions for autoloading and reloading
* enable and disable the Active Record query cache
* return acquired Active Record connections to the pool

> * constrain internal cache lifetimes

Reviewer: ```suggestion
* Track which threads are in safe positions for autoloading and reloading.
* Enable and disable the Active Record query cache.
* Return acquired Active Record connections to the pool.
* Constrain internal cache lifetimes.
```

end
```

NOTE: The Concurrent Ruby gem uses a `ThreadPoolExecutor`, which it sometimes

> configures with an `executor` option. Despite the name, it is unrelated to the

Reviewer: ```suggestion
configures with an `executor` option. Despite the name, it is _not_ related to the
```

end
```

NOTE: The Rails Executor is safely re-entrant; it can be called again if it is
already running. In this case, the `wrap` method has no effect.

> #### Running

Reviewer 2: Just "Running" is maybe a bit cryptic.
```suggestion
#### Running Mode
```

When called, the Rails Executor will put the current thread into `running` mode
in the [Load Interlock](#load-interlock).

This operation will block temporarily if another thread is currently either
autoloading a constant or unloading/reloading the application.

#### Examples of Wrapped Application Code

Any time your library or component needs to invoke code that will need to run in
the application, this code should be wrapped to ensure thread safety and a
consistent and clean runtime state.

For example, you may be setting a `Current` user (using
[`ActiveSupport::CurrentAttributes`](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html)).

```ruby
def log_with_user_context(message)
  Rails.application.executor.wrap do
    Current.user = User.find_by(id: 1)
  end
end
```

> You may be triggering an ActiveRecord callback or lifecycle hook in an

Reviewer 2: ```suggestion
You may be triggering an Active Record callback or lifecycle hook in an
```

application:

```ruby
def perform_task_with_record(record)
  Rails.application.executor.wrap do
    record.save! # Executes before_save, after_save, etc.
  end
end
```

Or enqueuing or performing a background job within the application:

```ruby
def enqueue_background_job(job_class, *args)
  Rails.application.executor.wrap do
    job_class.perform_later(*args)
  end
end
```

> These are just a few of many possible other use cases, including rendering views

Reviewer: ```suggestion
These are just a few of many possible use cases, including rendering views
```

or templates, broadcasting via [`Action Cable`](action_cable_overview.html) or
using [`Rails.cache`](caching_with_rails.html).

### The Reloader

Like the Executor, the
[Reloader](https://api.rubyonrails.org/classes/ActiveSupport/Reloader.html) also

> wraps application code. If the Executor is not already active on the current

Reviewer: The immediate question in my head is "okay so what is the difference between the Executor and the Reloader when?" The NOTE below answers that question. Maybe worth moving the text in that note up here?
